src/library_pipefs.js

```diff
@@ -0,0 +1,208 @@
+mergeInto(LibraryManager.library, {
+  $PIPEFS__postset: '__ATINIT__.push(function() { PIPEFS.root = FS.mount(PIPEFS, {}, null); });',
+  $PIPEFS__deps: ['$FS'],
+  $PIPEFS: {
+    BUCKET_BUFFER_SIZE: 1024 * 8, // 8KiB Buffer
+    mount: function (mount) {
+      return FS.createNode(null, '/', {{{ cDefine('S_IFDIR') }}} | 511 /* 0777 */, 0);
+    },
+    createPipe: function () {
+      var pipe = {
+        buckets: []
+      };
+
+      pipe.buckets.push({
+        buffer: new Uint8Array(PIPEFS.BUCKET_BUFFER_SIZE),
+        offset: 0,
+        roffset: 0
+      });
+
+      var rName = PIPEFS.nextname();
+      var wName = PIPEFS.nextname();
+      var rNode = FS.createNode(PIPEFS.root, rName, {{{ cDefine('S_IFIFO') }}}, 0);
+      var wNode = FS.createNode(PIPEFS.root, wName, {{{ cDefine('S_IFIFO') }}}, 0);
+
+      rNode.pipe = pipe;
+      wNode.pipe = pipe;
+
+      var readableStream = FS.createStream({
+        path: rName,
+        node: rNode,
+        flags: FS.modeStringToFlags('r'),
+        seekable: false,
+        stream_ops: PIPEFS.stream_ops
+      });
+      rNode.stream = readableStream;
+
+      var writableStream = FS.createStream({
+        path: wName,
+        node: wNode,
+        flags: FS.modeStringToFlags('w'),
+        seekable: false,
+        stream_ops: PIPEFS.stream_ops
+      });
+      wNode.stream = writableStream;
+
+      return {
+        readable_fd: readableStream.fd,
+        writable_fd: writableStream.fd
+      };
+    },
+    stream_ops: {
+      poll: function (stream) {
+        var pipe = stream.node.pipe;
+
+        if ((stream.flags & {{{ cDefine('O_ACCMODE') }}}) === {{{ cDefine('O_WRONLY') }}}) {
+          return ({{{ cDefine('POLLWRNORM') }}} | {{{ cDefine('POLLOUT') }}});
+        } else {
+          if (pipe.buckets.length > 0) {
+            for (var i = 0; i < pipe.buckets.length; i++) {
+              var bucket = pipe.buckets[i];
+              if (bucket.offset - bucket.roffset > 0) {
+                return ({{{ cDefine('POLLRDNORM') }}} | {{{ cDefine('POLLIN') }}});
+              }
+            }
+          }
+        }
+
+        return 0;
+      },
+      ioctl: function (stream, request, varargs) {
+        return ERRNO_CODES.EINVAL;
+      },
+      read: function (stream, buffer, offset, length, position /* ignored */) {
+        var pipe = stream.node.pipe;
+        var currentLength = 0;
+
+        pipe.buckets.forEach(function (bucket) {
+          currentLength += bucket.offset - bucket.roffset;
+        });
```

Comment:
I see there is a comment below about optimising to not generate extra garbage. In case that is interesting, this forEach construct also generates temporary garbage by creating an iterator object and a temporary function that gets nuked afterwards. The second iteration over the buckets below is using a `for (var i = 0; i < pipe.buckets.length; i++) {` construct. That is a nice and garbage-free way to iterate over an array.